libsolidity/analysis/ReferencesResolver.cpp

```diff
@@ -301,109 +301,66 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
-					typeLoc = DataLocation::CallData;
-				else
-					typeLoc = varLoc == Location::Memory ? DataLocation::Memory : DataLocation::Storage;
-			}
-			else if (_variable.isCallableParameter() && dynamic_cast<Declaration const&>(*_variable.scope()).isPublic())
-			{
-				auto const& contract = dynamic_cast<ContractDefinition const&>(
-					*dynamic_cast<Declaration const&>(*_variable.scope()).scope()
-				);
-				// force locations of public or external function (return) parameters to memory
-				if (varLoc != Location::Memory && varLoc != Location::Default && !contract.isLibrary())
-					fatalTypeError(_variable.location(),
-						"Location has to be memory for publicly visible functions "
-						"(remove the \"storage\" or \"calldata\" keyword)."
-					);
-				if (varLoc == Location::Default || !contract.isLibrary())
-					typeLoc = DataLocation::Memory;
 				else
-				{
-					if (varLoc == Location::CallData)
-						fatalTypeError(_variable.location(),
-							"Location cannot be calldata for non-external functions "
-							"(remove the \"calldata\" keyword)."
-						);
-					typeLoc = varLoc == Location::Memory ? DataLocation::Memory : DataLocation::Storage;
-				}
+					errorString += " for variable";
+				string varLocStr = _variable.locationToString(varLoc);
+				varLocStr = varLocStr == "none" ? "none" : "\"" + varLocStr + "\"";
+				errorString += ", but " + varLocStr + " was given.";
+				fatalTypeError(_variable.location(), errorString);
 			}
+
+			// Find correct data location.
+			if (_variable.isEventParameter() || _variable.isConstant())
+				typeLoc = DataLocation::Memory;
+			else if (_variable.isExternalCallableParameter() && varLoc == Location::Default)
+				typeLoc = DataLocation::CallData;
+			else if (varLoc == Location::Default)
+				typeLoc = DataLocation::Storage;
 			else
 			{
-				if (_variable.isConstant())
-				{
-					if (varLoc != Location::Default && varLoc != Location::Memory)
-						fatalTypeError(
-							_variable.location(),
-							"Data location has to be \"memory\" (or unspecified) for constants."
-						);
-					typeLoc = DataLocation::Memory;
-				}
-				else if (varLoc == Location::Default)
-				{
-					if (_variable.isCallableParameter())
-						typeLoc = DataLocation::Memory;
-					else
-					{
-						typeLoc = DataLocation::Storage;
-						if (_variable.isLocalVariable())
-							typeError(
-								_variable.location(),
-								"Data location must be specified as either \"memory\" or \"storage\"."
-							);
-					}
-				}
-				else
+				switch (varLoc)
 				{
-					switch (varLoc)
-					{
 					case Location::Memory:
 						typeLoc = DataLocation::Memory;
 						break;
 					case Location::Storage:
 						typeLoc = DataLocation::Storage;
 						break;
 					case Location::CallData:
-						fatalTypeError(_variable.location(),
-							"Variable cannot be declared as \"calldata\" (remove the \"calldata\" keyword)."
-						);
+						typeLoc = DataLocation::CallData;
 						break;
 					default:
```

Comment:
I think it would be better to just remove the default case, then we get an error from the compiler if we ever add new cases here.